an-introduction-to-unit-testing.md

```diff
@@ -0,0 +1,852 @@
+---
+title: An Introduction to Unit Testing
+domain: campzulu.hashnode.dev
+tags: Unit Testing, C#, NUnit
+cover: https://cdn.hashnode.com/res/hashnode/image/upload/v1668080970548/myC4ov8zN.jpeg
+publishAs: fabioscagliola
+ignorePost: true
+---
+
+# An Introduction to Unit Testing
+
+This article provides an introduction to unit testing.
+
+## Contents
+
+- Unit testing in theory
+- Unit testing in practice
+- Fakes
+- Code samples
+- Suggested readings
+
+## Unit testing in theory
+
+This section defines unit testing, units of work, integration tests, and test-driven development.
+
+### Unit testing definition 1
+
+A unit test is a piece of a code (usually a method) that invokes another piece of code and checks the correctness of some assumptions afterward. If the assumptions turn out to be wrong, the unit test has failed. The piece of code, also called the unit, is a method or function.
+
+### Unit of work
+
+A unit of work is the sum of actions that take place between the invocation of a public method in the system and a single noticeable end result after a test of that system.
+
+A noticeable end result can be observed without looking at the internal state of the system and only through its public APIs and behavior.
+
+### Unit test definition 2
+
+A unit test is a piece of code that invokes a unit of work and checks one specific end result of that unit of work. If the assumptions on the end result turn out to be wrong, the unit test has failed. The scope of a unit test can span as little as a method or as much as multiple classes.
+
+### Integration tests
+
+Integration tests are any tests that aren’t fast and consistent, and that use one or more real dependencies of the units under test. If the test uses the real system time, the real filesystem, a real database, or some piece of hardware, it has stepped into the realm of integration testing.
+
+### Integration tests drawbacks
+
+If a test uses the real database, then it’s no longer only running in memory; in that, its actions are harder to erase than when using only in-memory fake data. Integration tests increase the risk of another problem: testing too many things at once. In an integration test you can have many failure points and, therefore, finding the cause of the problem is harder. An integration test uses real dependencies.
```

Comment:
What is the recipe for this? Integration tests need to be executed in a well-defined, controlled sandbox.